WIP: Add support for ResponseHeaderModifier for HTTPRouteRule objects

kevin85421 opened the pull request:
### Proposed changes

Write a clear and concise description that helps reviewers understand the purpose and impact of your changes. Use the
following format:

Problem: Give a brief overview of the problem or feature being addressed.

Solution: Explain the approach you took to implement the solution, highlighting any significant design decisions or
considerations.

Testing: Describe any testing that you did.

Please focus on (optional): If you any specific areas where you would like reviewers to focus their attention or provide
specific feedback, add them here.

Closes #ISSUE

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [ ] I have read the [CONTRIBUTING](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that all unit tests pass after adding my changes
- [ ] I have updated necessary documentation
- [ ] I have rebased my branch onto main
- [ ] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

kevin85421 commented on internal/mode/static/nginx/config/http/config.go:
```diff
@@ -12,12 +12,13 @@ type Server struct {
+	ProxyPass           string
+	HTTPMatchVar        string
+	ProxySetHeaders     []Header
+	AddHeaderDirectives []Header
```
This PR uses the `add_header` directives to modify HTTP response headers. The variable name `AddHeaders` might be confused with `HTTPHeaderFilter.Add`. Hence, I use `AddHeaderDirectives` here.

kevin85421 commented on internal/mode/static/nginx/config/servers.go:
```diff
@@ -487,6 +489,37 @@ func generateProxySetHeaders(filters *dataplane.HTTPFilters) []http.Header {
 	return append(proxySetHeaders, headers...)
 }
 
+func generateAddHeaderDirectives(filters *dataplane.HTTPFilters) []http.Header {
```
Is there any relationship between response headers and `RequestURLRewrite`, similar to what is seen in `generateProxySetHeaders` ([code](https://github.com/nginxinc/nginx-gateway-fabric/blob/e521f06aed06d864f358fc20036375cdb7b4c2a3/internal/mode/static/nginx/config/servers.go#L454-L461))? This [link](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPURLRewriteFilter) said that "HTTPURLRewriteFilter defines a filter that modifies a request during forwarding.". Does it mean that `HTTPURLRewriteFilter` only affects requests?

kate-osborn commented:
@kevin85421 thanks for working on this! 

We have three actions for ResponseHeaderModifier: `add`, `set`,  and `remove`.  We also need to consider the cases where the header we are performing an action on already exists in the response headers. This gives us the following table:

| Case |Action | Header exists? | Expected behavior |
|-|-|-|-|
| 1 | add | no| set header to value |
| 2 | add | yes | append to existing header value |
| 3 | set | no | set header to value |
| 4 | set | yes | overwrite existing header value |
| 5 | remove | no | unset header (no-op) |
| 6 | remove | yes | unset header |

With your current implementation, cases 1, 2, and 3 work as expected, but the rest do not.

There are a couple of reasons for that. 

(1) In the manual test you described, you check the _request_ headers instead of the _response_  headers. More on testing later. 

(2) The `add_header` NGINX directive works differently from the `proxy_set_header` directive.
- It does not overwrite a header's value. 

  Instead, it adds another `header:value` pair. For example, say the upstream server responds with a response header `X-My-Header: existing-header`. The directive `add_header X-My-Header "new-header";` will add `X-My-Header: new-header` to the response headers, but it will not remove the `X-My-Header: existing-header` pair. 

- Similarly, when the value is `""`, it does not remove the header. It just doesn't add a new `header:value` pair to the response headers.

- It only adds a header when it receives a successful (2xx/3xx) response from the upstream server. This is easily fixed by adding the `always` parameter to the directive. `always` will ensure that NGINX always adds the header, regardless of the response code.


To get all the cases passing, we can add the following nginx directives:

(1) add:

```
add: 
- name: X-Header-Add
  value: header-add
```
==> 
```
 add_header_name X-Header-Add "header-add" always;
```

(2) set: we need two directives for this action:

```
set:
- name: X-Header-Set
  value: header-set
```
==> 
```
proxy_hide_header X-Header-Set;
add_header X-Header-Set "header-set" always;
```

(3) remove: 

```
remove:
- name: X-Header-Remove
```
==> 
```
proxy_hide_header X-Header-Remove;
```

There's no need for `map` in this solution because `add_header` does not overwrite the existing headers.

To verify the functionality, run the conformance tests locally. You will need to add `HTTPRouteResponseHeaderModifier` to the [list of supported features](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/conformance/Makefile#L6) and follow [the instructions](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/conformance/README.md) to run the tests. 

To iterate more quickly and run only the `HTTPRouteResponseHeaderModifier` test, you can add the go test flag `--run TestConformance/HTTPRouteResponseHeaderModifier ` to `go test` command in the `run-conformance-tests` Makefile target.

If you would like to test manually, ping me on Slack, and I can send you the Yaml files I used for testing. 

Thanks again for working on this!

kevin85421 commented:
@kate-osborn Thank you for your detailed reply!

>  In the manual test you described, you check the request headers instead of the response headers. More on testing later.

Do you mean that the following output from the `curl` command is still showing request headers instead of response headers?

```
< HTTP/1.1 200 OK
< Server: nginx/1.25.3
< Date: Mon, 22 Jan 2024 07:28:07 GMT
< Content-Type: text/plain
< Content-Length: 327
< Connection: keep-alive
< Accept-Encoding: response-accept-encoding
< My-cool-header: my-client-value,response-appended-value
< My-Overwrite-Header: response-this-is-the-only-value
< Host: echo.example.com:8080
< X-Forwarded-For: 127.0.0.1
< Connection: close
```

kate-osborn commented:
> @kate-osborn Thank you for your detailed reply!
> 
> > In the manual test you described, you check the request headers instead of the response headers. More on testing later.
> 
> Do you mean that the following output from the `curl` command is still showing request headers instead of response headers?
> 
> ```
> < HTTP/1.1 200 OK
> < Server: nginx/1.25.3
> < Date: Mon, 22 Jan 2024 07:28:07 GMT
> < Content-Type: text/plain
> < Content-Length: 327
> < Connection: keep-alive
> < Accept-Encoding: response-accept-encoding
> < My-cool-header: my-client-value,response-appended-value
> < My-Overwrite-Header: response-this-is-the-only-value
> < Host: echo.example.com:8080
> < X-Forwarded-For: 127.0.0.1
> < Connection: close
> ```

@kevin85421 sorry, I meant the output from the echo app:

```
Headers:
 header 'Accept-Encoding' is 'compress'
 header 'My-cool-header' is 'my-client-value,this-is-an-appended-value'
 header 'My-Overwrite-Header' is 'this-is-the-only-value'
 header 'Host' is 'echo.example.com:8080'
 header 'X-Forwarded-For' is '127.0.0.1'
 header 'Connection' is 'close'
 header 'Accept' is '*/*'
```

That app writes the request headers. 

The curl command you used does show the response headers.

kevin85421 commented:
I tried to test this PR manually. The `curl` command gets the following output which seems to fulfill the expected behaviors for all six cases mentioned in [this comment](https://github.com/nginxinc/nginx-gateway-fabric/pull/1494#issuecomment-1908626933). I will try to run the conformance test to figure out whether I have any misunderstanding about the expected behavior.

* Use this new [echo-route.yaml](https://gist.github.com/kevin85421/9cd765164a5b9233aacb703f7493626b).
  ```yaml
  filters:
      - type: ResponseHeaderModifier
        responseHeaderModifier:
          set:
          - name: To-Set-Exist-Header
            value: overwrite-value
          - name: Set-New-Header
            value: set-new-header
          add:
          - name: To-Add-Exist-Header
            value: response-appended-value
          - name: Add-New-Header
            value: add-new-header
          remove:
          - To-Remove-Exist-Header
          - Remove-Non-Exist-Header
  ```
* Submit a request with three headers, `To-Add-Exist-Header`, `To-Set-Exist-Header`, and `To-Remove-Exist-Header`.
  ```sh
  curl -v --resolve echo.example.com:$GW_PORT:$GW_IP http://echo.example.com:$GW_PORT/headers -H "To-Add-Exist-Header:original-add-value" -H "To-Set-Exist-Header:original-set-value" -H "To-Remove-Exist-Header:original-remove-value"
  ```
* The `curl` command gets the following output which seems to fulfill the expected behaviors for all six cases mentioned in [this comment](https://github.com/nginxinc/nginx-gateway-fabric/pull/1494#issuecomment-1908626933).
  <img width="535" alt="Screen Shot 2024-01-27 at 3 41 55 PM" src="https://github.com/nginxinc/nginx-gateway-fabric/assets/20109646/66e0c1f4-67bb-4b83-b6f8-04f532f9ce83">

kevin85421 commented:
I can't easily run conformance tests locally based on [this doc](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/conformance/README.md).

* Based on [HTTPRouteResponseHeaderModification](https://github.com/search?q=repo%3Akubernetes-sigs%2Fgateway-api+HTTPRouteResponseHeaderModification&type=code), we need to use `HTTPRouteResponseHeaderModification` instead of `HTTPRouteResponseHeaderModifier` for [SUPPORTED_FEATURES](https://github.com/nginxinc/nginx-gateway-fabric/blob/b77e9c2a42ba57f6228056d56a986eed4f531083/conformance/Makefile#L6), but I am not 100% sure.
* Use `--run-test TestConformance/HTTPRouteResponseHeaderModifier` instead of `--run`. I am also not 100% sure.
* Eventually, I chose not to use `--run-test`. Instead, I removed all features in `SUPPORTED_FEATURES` except for `HTTPRouteResponseHeaderModification`, i.e.
  ```sh
  SUPPORTED_FEATURES = HTTPRouteResponseHeaderModification
  ```
  * For tests in [httproute-response-header-modifier.go](https://github.com/kubernetes-sigs/gateway-api/blob/1edf7e93ae4bb8c5ea5cc2193586cfe1e5d15e2f/conformance/tests/httproute-response-header-modifier.go#L49), 3 tests pass and 5 tests fail.

kate-osborn commented:
@kevin85421

> Based on [HTTPRouteResponseHeaderModification](https://github.com/search?q=repo%3Akubernetes-sigs%2Fgateway-api+HTTPRouteResponseHeaderModification&type=code), we need to use HTTPRouteResponseHeaderModification instead of HTTPRouteResponseHeaderModifier for [SUPPORTED_FEATURES](https://github.com/nginxinc/nginx-gateway-fabric/blob/b77e9c2a42ba57f6228056d56a986eed4f531083/conformance/Makefile#L6), but I am not 100% sure.

Yes, you're right. I mistakenly gave you the test's ShortName instead of the SupportedFeature name.

> Use --run-test TestConformance/HTTPRouteResponseHeaderModifier instead of --run. I am also not 100% sure.

I was able to get the HTTPRouteResponseHeaderModifier test running by adding `--run TestConformance/HTTPRouteResponseHeaderModifier` to the `go test` command in the conformance Makefile:

`go test -v . -run TestConformance/HTTPRouteResponseHeaderModifier -tags conformance,experimental...`

I haven't tried using the `-run-test` argument, but seems like that works as well.

> The [doc](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/conformance/README.md) installs NGF in provisioning mode instead of static mode. I am not sure what's the difference between these two modes.

We use this mode only for conformance tests because the tests create multiple Gateway resources, which the `static` mode does not support. The`provisioner` mode runs [this code](https://github.com/nginxinc/nginx-gateway-fabric/tree/main/internal/mode/provisioner) which implements data plane provisioning for NGINX Gateway Fabric (NGF): it creates an NGF static mode Deployment for each Gateway that belongs to the provisioner GatewayClass.

> After the conformance tests were complete (fail), I tried to submit curl requests to reproduce the 5 failed tests, but the requests couldn't be sent to the backend servers correctly. The Gateway same-namespace is in the GatewayConflict status, and I am not sure whether it is expected or not.

The conformance tests clean up after themselves, so my guess is that all the resources were deleted after the tests finished, which is why you weren't able to reach any of the backends.

> The Gateway same-namespace is in the GatewayConflict status, and I am not sure whether it is expected or not.

That's not expected. Were you running NGF in `provisioner` mode? When I ran the conformance tests on your branch using the `make install-ngf-local-build` and then `make run-conformance-tests`, 3/8 of the test cases for the HTTPRouteResponseHeaderModification passed, and all the Gateways were `Programmed`.

![Screen Shot 2024-01-29 at 3 06 26 PM](https://github.com/nginxinc/nginx-gateway-fabric/assets/50597707/75ec4a35-8717-45b7-b788-c3c0daf4a4ba)

> I am not sure why there are no same headers in https://github.com/nginxinc/nginx-gateway-fabric/pull/1494#issuecomment-1913370626.

The HTTPRouteResponseHeaderModifier test uses this [echo-server](https://github.com/kubernetes-sigs/ingress-controller-conformance/tree/master/images/echoserver) as the Backend, which allows you to set the _response_ headers for a request by specifying them in the `X-Echo-Set-Header` _request_ header. The echo-server reads in the header key-value pairs from the `X-Echo-Set-Header` _request_ header and writes each key-value pair as _response_ headers (see this [function](https://github.com/kubernetes-sigs/ingress-controller-conformance/blob/d920ed36a0076e169a9a329a850844ab3a695ae8/images/echoserver/echoserver.go#L168)).

When you sent the curl request `curl -v http://127.0.0.1:8080/set -H "X-Echo-Set-Header:Some-Other-Header:val,X-Header-Set:some-other-value"` you essentially told the echo-server to set the following _response_ headers: `Some-Other-Header: val` and `X-Header-Set: some-other-value`. 

Before, when you were testing with our [header example backend](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/examples/http-header-filter/headers.yaml), you were not setting the _response_ headers in your curl request. Instead, you were setting the _request_ headers. Our header example backend is not as smart as the echo-server, and there is no way to set the _response_ headers with it. 


> The curl command gets the following output which seems to fulfill the expected behaviors for all six cases mentioned in https://github.com/nginxinc/nginx-gateway-fabric/pull/1494#issuecomment-1908626933.

When verifying response headers, add the `-i` flag to your curl commands. From the curl man page:

```
 -i, --include
              Include the HTTP response headers in the output. The HTTP response headers can include things like server name, cookies, date of the document, HTTP version and more...

              To view the request headers, consider the -v, --verbose option.
```

I thought `-v` included both request and response headers, but the man page tells a different story.



I hope this helps, thanks again for your effort on this! Feel free to ping me on slack if you run into any more issues.

kevin85421 commented on internal/mode/static/nginx/config/servers.go:
```diff
@@ -487,6 +493,50 @@ func generateProxySetHeaders(filters *dataplane.HTTPFilters) []http.Header {
 }
 
+func generateAddResponseHeaders(filters *dataplane.HTTPFilters) []http.Header {
+	headers := make([]http.Header, len(baseHeaders))
```
This function also handles `baseHeaders`.

kate-osborn replied:
I don't think we need to handle `baseHeaders` in this function. The `baseHeaders` are request headers and are handled by `generateProxySetHeaders`.

kevin85421 replied:
Update https://github.com/nginxinc/nginx-gateway-fabric/pull/1494/commits/bcfa113486f0f38caa938a3051a8064fed3908a7

kevin85421 commented on internal/mode/static/nginx/config/servers.go:
```diff
@@ -487,6 +493,50 @@ func generateProxySetHeaders(filters *dataplane.HTTPFilters) []http.Header {
+	responseAddHeaders := make([]http.Header, 0, headerLen)
+	if len(headerFilter.Add) > 0 {
+		// FIXME(kevin85421): Should we use a different function?
+		addHeaders := convertSetHeaders(headerFilter.Add)
```
Should we rename the function or create a new function?

kate-osborn replied:
Let's rename. 

`convertSetHeaders` is just converting  `[]dataplane.HTTPHeader` to `[]http.Header`.  Let's name it something that describes that behavior.

I think `convertAddHeaders`  can be renamed as well. It does a similar thing as `convertSetHeaders` except it generates a map variable name.

kevin85421 replied:
Do you have any preferred function name?

kate-osborn replied:
Naming is hard 😅 Maybe `createHeaders` and `createHeadersWithVarName`? Open to other names though

kevin85421 replied:
Update https://github.com/nginxinc/nginx-gateway-fabric/pull/1494/commits/d316a416f50524fd4de0dbb318d3c864e543b835 https://github.com/nginxinc/nginx-gateway-fabric/pull/1494/commits/9bf3447aecf8594e359ae07a61971e09adf64323

kevin85421 commented on internal/mode/static/nginx/config/http/config.go:
```diff
@@ -12,12 +12,15 @@ type Server struct {
+	ProxyPass             string
+	HTTPMatchVar          string
+	ProxySetHeaders       []Header
+	AddResponseHeaders    []Header
```
We might consider reusing `dataplane.HTTPHeaderFilter` instead of using `AddResponseHeaders`, `SetResponseHeaders`, and `RemoveResponseHeaders`. Why do we need both `Header` and `dataplane.HTTPHeader`?

kevin85421 commented:
> Use --run-test TestConformance/HTTPRouteResponseHeaderModifier instead of --run. I am also not 100% sure.

I will get the following error when I use `--run`.

```sh
kubectl run -i conformance \
      --image=conformance-test-runner:latest --image-pull-policy=Never \
        --overrides='{ "spec": { "serviceAccountName": "conformance" }  }' \
        --restart=Never -- sh -c "go test -v . -tags conformance,experimental -args --gateway-class=nginx \
                                                        --run TestConformance/HTTPRouteResponseHeaderModifier \
                                                --supported-features=HTTPRouteResponseHeaderModification --version= \
                                                        --report-output=output.txt; cat output.txt" | tee output.txt
If you don't see a command prompt, try pressing enter.
flag provided but not defined: -run
```

> Were you running NGF in provisioner mode?

I installed NGF in static mode instead of running `make install-ngf-local-build`.

kevin85421 commented:
Manual test
```sh
make install-ngf-local-build
make run-conformance-tests
# Test passes. See https://gist.github.com/kevin85421/f9aa82a311a52e1240c70ae3f088f722 for more details.
```

kevin85421 commented:
> Instead, it adds another header:value pair. For example, say the upstream server responds with a response header X-My-Header: existing-header. The directive add_header X-My-Header "new-header"; will add X-My-Header: new-header to the response headers, but it will not remove the X-My-Header: existing-header pair.

I followed your suggestions and removed `map` from the `add` headers, and the conformance tests pass. However, I don't really understand the expected behavior of `add`. Take [this test](https://github.com/kubernetes-sigs/gateway-api/blob/1edf7e93ae4bb8c5ea5cc2193586cfe1e5d15e2f/conformance/tests/httproute-response-header-modifier.go#L95-L111) as an example, the backend will set `Some-Other-Header:val` and `X-Header-Add:some-other-value`, and the `ResponseHeaderModifier` for the path `/add` is shown as follow.

```yaml
    filters:
    - type: ResponseHeaderModifier
      responseHeaderModifier:
        add:
        - name: X-Header-Add
          value: add-appends-values
```

In my expectation, the response headers should be `Some-Other-Header:val` and `X-Header-Add:some-other-value,add-appends-values`. However, when I tried to reproduce the test manually by running the following `curl` command.  The response headers have `Some-Other-Header:val`, `X-Header-Add:some-other-value`, and `X-Header-Add:add-appends-values`. I am not sure why conformance tests pass when there are two `X-Header-Add` headers.

```sh
curl -i http://127.0.0.1:8080/add -H "X-Echo-Set-Header:Some-Other-Header:val,X-Header-Add:some-other-value"
```

kate-osborn commented:
> > Use --run-test TestConformance/HTTPRouteResponseHeaderModifier instead of --run. I am also not 100% sure.
> 
> I will get the following error when I use `--run`.
> 
> ```shell
> kubectl run -i conformance \
>       --image=conformance-test-runner:latest --image-pull-policy=Never \
>         --overrides='{ "spec": { "serviceAccountName": "conformance" }  }' \
>         --restart=Never -- sh -c "go test -v . -tags conformance,experimental -args --gateway-class=nginx \
>                                                         --run TestConformance/HTTPRouteResponseHeaderModifier \
>                                                 --supported-features=HTTPRouteResponseHeaderModification --version= \
>                                                         --report-output=output.txt; cat output.txt" | tee output.txt
> If you don't see a command prompt, try pressing enter.
> flag provided but not defined: -run
> ```


`--run` is an argument for `go test`, so it has to be placed before `-args`. The arguments after `-args` are passed to the test binary.

kevin85421 commented:
> `--run` is an argument for `go test`, so it has to be placed before `-args`. The arguments after `-args` are passed to the test binary.

I feel so embarrassed about my dumb question lol.

kate-osborn commented:
> > Instead, it adds another header:value pair. For example, say the upstream server responds with a response header X-My-Header: existing-header. The directive add_header X-My-Header "new-header"; will add X-My-Header: new-header to the response headers, but it will not remove the X-My-Header: existing-header pair.
> 
> I followed your suggestions and removed `map` from the `add` headers, and the conformance tests pass. However, I don't really understand the expected behavior of `add`. Take [this test](https://github.com/kubernetes-sigs/gateway-api/blob/1edf7e93ae4bb8c5ea5cc2193586cfe1e5d15e2f/conformance/tests/httproute-response-header-modifier.go#L95-L111) as an example, the backend will set `Some-Other-Header:val` and `X-Header-Add:some-other-value`, and the `ResponseHeaderModifier` for the path `/add` is shown as follow.
> 
> ```yaml
>     filters:
>     - type: ResponseHeaderModifier
>       responseHeaderModifier:
>         add:
>         - name: X-Header-Add
>           value: add-appends-values
> ```
> 
> In my expectation, the response headers should be `Some-Other-Header:val` and `X-Header-Add:some-other-value,add-appends-values`. However, when I tried to reproduce the test manually by running the following `curl` command. The response headers have `Some-Other-Header:val`, `X-Header-Add:some-other-value`, and `X-Header-Add:add-appends-values`. I am not sure why conformance tests pass when there are two `X-Header-Add` headers.
> 
> ```shell
> curl -i http://127.0.0.1:8080/add -H "X-Echo-Set-Header:Some-Other-Header:val,X-Header-Add:some-other-value"
> ```

@kevin85421 I believe that multiple entries of the same header name are equivalent to one header name with a comma-separated list of values, so long as the order of the headers is correct. To be sure, I asked this question in the Gateway API slack channel: https://kubernetes.slack.com/archives/CR0H13KGA/p1707260286602689.

kate-osborn commented:
> @kevin85421 I believe that multiple entries of the same header name are equivalent to one header name with a comma-separated list of values, so long as the order of the headers is correct. To be sure, I asked this question in the Gateway API slack channel: https://kubernetes.slack.com/archives/CR0H13KGA/p1707260286602689.

@kevin85421 confirmed that they are equivalent. I hope that answers your question. I will take a look at the rest of your PR this week 🙂

kevin85421 commented:
@kate-osborn Thank you!

kate-osborn commented on internal/mode/static/nginx/config/http/config.go:
```diff
@@ -12,12 +12,15 @@ type Server struct {
+	ProxyPass             string
+	HTTPMatchVar          string
+	ProxySetHeaders       []Header
+	AddResponseHeaders    []Header
```
Let's introduce a struct for response headers:

```
type ResponseHeaders struct {
    Add []Header
    Set []Header
    Remove []Header
}
```

It reduces the number of fields in Location and I think it's a bit easier to read.

kevin85421 replied:
Updated https://github.com/nginxinc/nginx-gateway-fabric/pull/1494/commits/629ee127585f277e3e41eafc1d47fd86ea583c43.

kate-osborn commented on internal/mode/static/nginx/config/servers.go:
```diff
@@ -247,6 +247,9 @@ func updateLocationsForFilters(
 
 	rewrites := createRewritesValForRewriteFilter(filters.RequestURLRewrite, path)
 	proxySetHeaders := generateProxySetHeaders(&matchRule.Filters)
+	addResponseHeaders := generateAddResponseHeaders(&matchRule.Filters)
```
Another bonus of introducing a `ResponseHeaders` struct is we can collapse these three lines into one:

```
responseHeaders := generateResponseHeaders(&matchRule.Filters)
```

Plus, then we only have to handle this case `if filters == nil || filters.ResponseHeaderModifiers == nil {
		return []http.Header{}
	}` once.

kevin85421 replied:
Update https://github.com/nginxinc/nginx-gateway-fabric/commit/629ee127585f277e3e41eafc1d47fd86ea583c43 https://github.com/nginxinc/nginx-gateway-fabric/pull/1494/commits/d8b35c0ceeb14c45b416469dc6e7480a04ff1329

kate-osborn commented on internal/mode/static/nginx/config/servers.go:
```diff
@@ -487,6 +493,50 @@ func generateProxySetHeaders(filters *dataplane.HTTPFilters) []http.Header {
+		return []http.Header{}
+	}
+	headerFilter := filters.ResponseHeaderModifiers
+	headerLen := len(headerFilter.Set)
```
nit: no need to declare a variable here, just access the length in the `make` function.

kevin85421 replied:
Update https://github.com/nginxinc/nginx-gateway-fabric/pull/1494/commits/95e60390f78674d32e283630578bf3cbb922d137.

kate-osborn commented on internal/mode/static/nginx/config/servers.go:
```diff
@@ -487,6 +493,50 @@ func generateProxySetHeaders(filters *dataplane.HTTPFilters) []http.Header {
+	removeHeaders := filters.ResponseHeaderModifiers.Remove
+	headerLen := len(removeHeaders)
+	responseRemoveHeaders := make([]string, headerLen)
+	copy(responseRemoveHeaders, removeHeaders)
```
Is this necessary? Can't we just return `filters.RequestHeaderModifiers.Remove` since it's a `[]string`?

kevin85421 replied:
We need to make a deep copy to prevent the slice from being accidentally modified.

```go
import (
	"fmt"
)

// processStrings takes a slice of strings as input and returns a new slice of strings
// where "-processed" is appended to each input string.
func processStrings(input []string) []string {
	return input
}

func main() {
	inputSlice := []string{"apple", "banana", "cherry"}
	outputSlice := processStrings(inputSlice)
	outputSlice[0] = "haha"
	fmt.Println("Input Slice:", inputSlice) // Input Slice: [haha banana cherry]
	fmt.Println("Output Slice:", outputSlice) // Output Slice: [haha banana cherry]
}
```

kate-osborn replied:
Good point. Can you add a comment in the code describing this?

kevin85421 replied:
Update https://github.com/nginxinc/nginx-gateway-fabric/pull/1494/commits/9fd505084b3185db74a9904fff3ac5324b797e68

kate-osborn commented on internal/mode/static/state/graph/httproute.go:
```diff
@@ -806,16 +809,16 @@ func validateFilterRewrite(
 }
 
 func validateFilterHeaderModifier(
+	filterType v1.HTTPRouteFilterType,
```
Do we need to pass the filterType in? Can we just add the filter type to the path when we pass it to function:

```
validateFilterHeaderModifier(validator, filter.ResponseHeaderModifier, filterPath.Child("responseHeaderModifier"))
```

kevin85421 replied:
Is there any elegant way to retrieve the child of the `filterPath` (i.e. `"responseHeaderModifier"`) so that I can use it in `panicForBrokenWebhookAssumption(errors.New(filterTypeStr + " cannot be nil"))`?

kevin85421 replied:
Updated https://github.com/nginxinc/nginx-gateway-fabric/pull/1494/commits/2213a64d0182350664711d667ad00914995716e6. Note that `string(filter.Type)` is `RequestHeaderModifier` instead of `requestHeaderModifier`. I am not sure whether the capitalization of the first character matters or not.

kate-osborn reviewed (changes requested):
@kevin85421 

Left some suggestions, but overall the approach looks good! There may be additional validation work around disallowing special header names, but I'm still trying to figure that out. I will get back to you once I've had a chance to talk to the other maintainers. 

There's also some docs changes that are needed with this work.

You'll need to update the [compatibility doc](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/site/content/overview/gateway-api-compatibility.md) to state that we support this filter type.

The [issue](https://github.com/nginxinc/nginx-gateway-fabric/issues/1397) also asks for a how-to guide that describes how to use this filter. Examples of these guides can be found [here](https://github.com/nginxinc/nginx-gateway-fabric/tree/main/site/content/how-to/traffic-management). How-to guides are a lot of work, so I'm happy to split that out into a separate issue if you rather not tackle that as part of this PR. 

Finally, once you mark the PR as ready, make sure to add the response header filter feature to the list of supported features in the conformance Makefile 😄

kevin85421 commented:
@kate-osborn I have already addressed almost all comments. Thank you for the review!

* https://github.com/nginxinc/nginx-gateway-fabric/pull/1494#pullrequestreview-1881511615
  * [Wait] There may be additional validation work around disallowing special header names, but I'm still trying to figure that out.
  * [Done] You'll need to update the [compatibility doc](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/site/content/overview/gateway-api-compatibility.md) to state that we support this filter type.
  * [TODO: Follow up PR]: how-to guide that describes how to use this filter.
  * [Done]: make sure to add the response header filter feature to the list of supported features in the conformance Makefile
* [Wait] https://github.com/nginxinc/nginx-gateway-fabric/pull/1494#discussion_r1490226060 

Manual test

```sh
make install-ngf-local-build
make run-conformance-tests
# conformance-profile.yaml: https://gist.github.com/kevin85421/d11309cbd4b1245ad4acba40d4c0843b
```

kevin85421 commented:
Btw, `gitleaks` complains about this [fake private key](https://github.com/nginxinc/nginx-gateway-fabric/blob/6becde2d07d89a9cbb83e623dddf7488ea171f81/internal/mode/static/state/graph/configmaps_test.go#L35) for testing, so I need to run `SKIP=gitleaks git commit` to commit.

ADubhlaoich reviewed (commented):
LGTM for the docs changes: I am not approving because most of the PR is code-based.

kate-osborn commented:
> Btw, `gitleaks` complains about this [fake private key](https://github.com/nginxinc/nginx-gateway-fabric/blob/6becde2d07d89a9cbb83e623dddf7488ea171f81/internal/mode/static/state/graph/configmaps_test.go#L35) for testing, so I need to run `SKIP=gitleaks git commit` to commit.

🤔 this key is in the `.gitleaksignore` file: https://github.com/nginxinc/nginx-gateway-fabric/blob/a66255b593c4cdb7aaa1b340ab425cc27c90aa37/.gitleaksignore#L6.

Perhaps rebasing will solve the issue?

kate-osborn reviewed (changes requested):
Following up on extra validation:

Let's disallow the following special header fields in the ResponseHeaderModifier:
- "Server"
- "Date"
- "X-Pad"
- Headers starting with "X-Accel"
- "Content-Type"
- "Content-Length"
- "Connection"

After some testing, trying to modify these headers in the same way we modify other headers doesn't work. I think it's better to disallow them through validation and document in the compatibility doc that they aren't supported header values.

kevin85421 commented on internal/mode/static/nginx/config/servers.go:
```diff
@@ -535,7 +537,38 @@ func generateProxySetHeaders(filters *dataplane.HTTPFilters) []http.Header {
-func convertAddHeaders(headers []dataplane.HTTPHeader) []http.Header {
+func generateResponseHeaders(filters *dataplane.HTTPFilters) http.ResponseHeaders {
+	if filters == nil || filters.ResponseHeaderModifiers == nil {
+		return http.ResponseHeaders{
```
Updated https://github.com/nginxinc/nginx-gateway-fabric/pull/1494/commits/eef622444adada948dd36195c9e7156a3a876c45

kevin85421 commented on internal/mode/static/nginx/config/servers.go:
```diff
@@ -535,7 +537,38 @@ func generateProxySetHeaders(filters *dataplane.HTTPFilters) []http.Header {
+	responseRemoveHeaders := make([]string, len(headerFilter.Remove))
+	if len(headerFilter.Add) > 0 {
+		addHeaders := createHeaders(headerFilter.Add)
+		responseAddHeaders = append(responseAddHeaders, addHeaders...)
```
Updated https://github.com/nginxinc/nginx-gateway-fabric/pull/1494/commits/a3eaa90b1f7781664f1950e4c7854f327eb0a07b

kevin85421 commented on internal/mode/static/nginx/config/servers.go:
```diff
@@ -535,7 +537,38 @@ func generateProxySetHeaders(filters *dataplane.HTTPFilters) []http.Header {
+	headerFilter := filters.ResponseHeaderModifiers
+	responseAddHeaders := make([]http.Header, 0, len(headerFilter.Add))
+	responseSetHeaders := make([]http.Header, 0, len(headerFilter.Set))
+	responseRemoveHeaders := make([]string, len(headerFilter.Remove))
```
Updated https://github.com/nginxinc/nginx-gateway-fabric/pull/1494/commits/a3eaa90b1f7781664f1950e4c7854f327eb0a07b

kevin85421 commented on site/content/overview/gateway-api-compatibility.md:
```diff
@@ -157,7 +157,8 @@ See the [static-mode]({{< relref "/reference/cli-help.md#static-mode">}}) comman
       - `requestHeaderModifier`: Supported. If multiple filters are configured, NGINX Gateway Fabric will choose the first and ignore the rest. Incompatible with `urlRewrite`.
       - `urlRewrite`: Supported. If multiple filters are configured, NGINX Gateway Fabric will choose the first and ignore the rest. Incompatible with `requestHeaderModifier`.
-      - `responseHeaderModifier`, `requestMirror`, `extensionRef`: Not supported.
+      - `responseHeaderModifier`: Supported.
```
Updated https://github.com/nginxinc/nginx-gateway-fabric/pull/1494/commits/ddfd470ddba907d62ef0ed0fe6c39a9cdad69fdc

kevin85421 commented:
Thank @ADubhlaoich and @kate-osborn for the review! I have already addressed some comments. I have been very busy recently. I will do my best to address https://github.com/nginxinc/nginx-gateway-fabric/pull/1494#pullrequestreview-1891609353 and https://github.com/nginxinc/nginx-gateway-fabric/pull/1494#discussion_r1496545569 as soon as possible, possibly by next weekend.

kevin85421 commented:
I will update this PR today.

kevin85421 commented on internal/mode/static/state/graph/httproute.go:
```diff
@@ -879,6 +880,61 @@ func validateFilterHeaderModifierFields(
+	responseHeaderModifier *v1.HTTPHeaderFilter,
+	filterPath *field.Path,
+) field.ErrorList {
+	if errList := validateFilterHeaderModifier(validator, responseHeaderModifier, filterPath); errList != nil {
```
If `responseHeaderModifier` is invalid for both `validateFilterHeaderModifier` and `disallowedResponseHeaderSet`, should we include both in `allErrs`, or do we only need to return `validateFilterHeaderModifier`?

kate-osborn replied:
I think we only need to return the errors from `validateFilterHeaderModifier`.

@pleshakov, does that seem right to you?

pleshakov replied:
> I think we only need to return the errors from validateFilterHeaderModifier.

Sounds reasonable.

In the check, I would check the length of errList, not that it doesn't equal to nil - just so that we don't dependent on zero-length vs nil slice differences.

kevin85421 commented:
* Add a validation function based on https://github.com/nginxinc/nginx-gateway-fabric/pull/1494#pullrequestreview-1891609353: https://github.com/nginxinc/nginx-gateway-fabric/pull/1494/commits/12123fb1258046d8fa02e6c9e00b806d146dce08
  * TODO: Update docs
  * TODO: Try testing the validation function manually. Will NGF log error messages or file Kubernetes events if the header modifier is invalid?

I will update the docs after https://github.com/nginxinc/nginx-gateway-fabric/pull/1494#discussion_r1519713368 got answered.

* TODO: https://github.com/nginxinc/nginx-gateway-fabric/pull/1494#discussion_r1496545569
